Fix deletion of selected files in trashbin

sharidas opened the pull request:
When files were selected all, and later
selective files were opted for deletion,
the deletion was not happening. The reason
for this behaviour is filenames are passed
through ajax as seen in ui. Where as that
is not the case. This change helps to fix it.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
The filenames passed in the ajax were the files seen in UI. But that is not the case with trashbin files. This change fixes the issue.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/owncloud/core/issues/30942

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
The filenames passed in the ajax were the files seen in UI. But that is not the case with trashbin files. This change fixes the issue.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- [x] Create 2 files. And delete them.
- [x] Select all and then change the decision to select only one. Delete the file. The file is deleted.
- [x] The same experiment is repeated with multiple files. And the files were deleted.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

PVince81 commented on apps/files_trashbin/js/filelist.js:
```diff
@@ -215,6 +215,19 @@
+				/*
+				The getSelectedFiles gives the file name as it is seen in UI.
+				But the fact is filenames are storead as filename.dxxxxx. This
+				change helps to extract exact filename.
```
the filenames are stored like this because you can have multiple files in trashbin with the same name from different deletions. we need to use the suffixed name when talking to the server so I doubt that this approach can work correctly

sharidas commented on apps/files_trashbin/js/filelist.js:
```diff
@@ -215,6 +215,19 @@
+					filesFromAttribute.push(trSelectedFiles[i].getAttribute('data-file'));
+				}
+				if (files !== filesFromAttribute) {
+					files = filesFromAttribute;
```
@PVince81 This change gives the suffixed name i.e if the file name is `a.txt`, then the data passed to server will be of the form `a.txt.dxxxxxxx`. While testing with single and multiple files, I was able to see files getting deleted from trashbin. Also verified files getting deleted from `files_trashbin/files`. Yes you are right, the issue was caused because suffixed filenames were not passed to the server.

PVince81 replied:
I see, makes sense.

PVince81 commented:
next up is adjusting unit tests.

you can run them with `make test-js-debug` and if you need the browser for stepping with a debugger, open "http://localhost:9876/debug.html" (white page) and open the browser console

phil-davis commented:
Selected delete in trashbin works with this code (manual test).

Selected restore does not work. I expect the same sort of code changes are needed for the code that runs when the "Restore" button is clicked.

sharidas commented:
I have updated the change for restore issue. Will be looking into the unit tests.

phil-davis commented:
@sharidas webUI tests for this are available. See the 2nd commit in #31024 

The 2 new scenarios fail against current master, and pass when your code is included - good.

Please include those tests in this PR. You will need to rebase this PR up to current master (to get the recent webUI test methods that are used). Then cherry-pick that commit c736a0726b48277973189f75bb15dafa160ab90d into this PR and push it up to GitHub.

phil-davis commented:
I realized that I can add the test commit myself to this PR, so I have done that.

phil-davis commented:
@PVince81 unit tests and webUI acceptance tests have been added here.
Ready for review.

PVince81 commented on apps/files_trashbin/js/filelist.js:
```diff
@@ -172,6 +172,19 @@
+				for (var i = 0; i < trSelectedFiles.length; i++) {
+					filesFromAttribute.push(trSelectedFiles[i].getAttribute('data-file'));
+				}
+				if (files !== filesFromAttribute) {
```
I'm not sure this comparison is accurate as it might be comparing only the pointer reference and not the contents. Please use a clearer approach for this check.

phil-davis replied:
Hmmm - I took the test above, copy-paste, and modified the clicking at the start and expected results. Now I have to actually learn how to do this.

PVince81 commented on apps/files_trashbin/js/filelist.js:
```diff
@@ -215,6 +228,19 @@
+				for (var i = 0; i < trSelectedFiles.length; i++) {
+					filesFromAttribute.push(trSelectedFiles[i].getAttribute('data-file'));
+				}
+				if (files !== filesFromAttribute) {
```
duplicate code from above, please move to common method

phil-davis commented:
@PVince81 I refactored that. And also, ``_onClickDeleteSelected`` and ``_onClickRestoreSelected`` are specifically in the trashbin JS code, and they are only called to work on selected files, so they can always work on the ``data-file`` attribute. There is no need to ever do ``_.pluck(this.getSelectedFiles(), 'name')`` and so that removes the need for any comparison of the 2 things (that you commented above.

sharidas commented:
@phil-davis Even I was having same thought when asked to move the changes to common function. The `_pluck(this.getSelectedFiles(), 'name'` can be safely removed and no more comparison required.

PVince81 commented on apps/files_trashbin/js/filelist.js:
```diff
@@ -171,7 +186,7 @@
 				};
 			}
 			else {
-				files = _.pluck(this.getSelectedFiles(), 'name');
```
Now that I understand this PR better, I'm a bit puzzled that the "name" attribute doesn't contain the combined name because the name displayed in the UI is supposed to be "displayName". Maybe that's another bug or regression.

An alternative and potentially better fix would be to adjust this class' `elementToFile()` to add another attribute on the object containing the combined name.

phil-davis replied:
I did acceptance and JS tests here, and just did a bit of refactoring of the JavaScript code because I thought it looked easy enough and would get this PR finished.

Now that there are classes and attributes and objects to be massaged, I will leave it to someone with real JavaScript coding skills :) @sharidas ?

At least there is a good set of JS unit tests and webUI acceptance tests, so you now easily know if a refactoring breaks some "important" behavior.

sharidas replied:
The problem happens when user select all and the unselect(s) from the list. In my case I tried select all. 
1) The files I had in the trash were `a.txt` and `b.txt`. Now I unselected `b.txt`. Next tried to delete `a.txt`. Below is the screenshot where name has only a.txt:
![notworking](https://user-images.githubusercontent.com/3600427/38430237-55e4cf4e-39de-11e8-93ce-2c87321e9874.png)

2) Without selecting select all, and if I select specifically a file: `a.txt` the name becomes:
![working](https://user-images.githubusercontent.com/3600427/38430282-789b5698-39de-11e8-8c26-c6ed35e4f68d.png)

I guess @PVince81 the second option, right?

PVince81 commented on tests/acceptance/features/webUITrashbin/restore.feature:
```diff
@@ -50,6 +50,29 @@ So that I can recover accidentally deleted files/folders in ownCloud
 		But the file "data.zip" should not be listed in the files page on the webUI
 		And the folder "simple-folder" should not be listed in the files page on the webUI
 
+	Scenario: Select all except for some trashbin files and restore them in a batch
```
@phil-davis can you add a case where you delete two entries with the same name ?
basically create a folder "test", then deleted it, then another "test", delete it. they'd appear as two entries.
internally they have names with the timestamp appended (test.dXXXXXXX) but visually they look the same

sharidas commented:
Searched till 9.1.1 version. And this is broken in version 9.1.1.

PVince81 commented:
so version 9.1.0 was fine ? any changes ? git bisect will tell you

PVince81 commented:
Introducing "realname" is likely just a workaround for the real issue.
I suspect that the real issue is inside of `getSelectedFiles()` and the selection logic. So while it might work it around here it might defer the issue to a later point.

sharidas commented:
This is my understanding from the code, the variable `_selectedFiles` gets populated when:
- page is loaded ( page of the file list )
- when select all is done
The `_selectedFiles` populates the data under these circumstances, which is then used by `getSelectedFiles`. So lets say if the user deletes `a.txt` and `b.txt`, the files listed in the trashbin ui will be of same name. Instead of `a.txt.dxxxxx` and `b.txt.dxxxx` as in the filesystem. So in the new approach, what I have done is during select all, tried to update the files data which is used to populate `_selectedFiles`.

PVince81 commented on apps/files/js/filelist.js:
```diff
@@ -692,8 +692,26 @@
 				.closest('tr').toggleClass('selected', checked);
 			this._selectedFiles = {};
 			this._selectionSummary.clear();
+			var getAllSelected = this.$el.find('tr.selected');
```
you won't receive them all. Remember that when there are more than 20+ items on the page, only the first items are rendered. You cannot rely on the DOM. Rely on "this.files" to have the full list.

sharidas commented on apps/files/js/filelist.js:
```diff
@@ -694,6 +694,13 @@
+					If the selection is from trashbin we need to do update the
+					name of the files.
+					 */
+					if (this.$fileList.selector === "#app-content-trashbin #fileList") {
```
A check made to figure out if the selection happens from trashbin UI. Not sure how to improvise this piece of comparison here. Intention is that no other ui should get updated name in `this.files`.

PVince81 replied:
filelist from apps/files should not know anything about apps/files_trashbin, so we need to find a way to make it work without violating this

sharidas commented on apps/files/js/filelist.js:
```diff
@@ -694,6 +694,13 @@
+					name of the files.
+					 */
+					if (this.$fileList.selector === "#app-content-trashbin #fileList") {
+						this.files[i].name = this.files[i].name + '.d' + this.files[i].mtime/1000;
```
The division with 1000 was borrowed after having a look at the logic here: https://github.com/owncloud/core/blob/master/apps/files/lib/Helper.php#L142

phil-davis commented:
@sharidas @PVince81 rebased and conflict resolved.

This works and has tests. So where do we go from here with the discussion about the actual code?

PVince81 commented:
The current solution is a hack as the base FileList should not know anything about trashbin as it would be a dependency between the "files" app and "files_trashbin" app. This would introduce some technical debt.

phil-davis commented:
So is "somebody" = @sharidas ? designing a non-hacky way to fix this?

It seems a shame to have a working and tested fix, but not be able to release it.

sharidas commented:
> It seems a shame to have a working and tested fix, but not be able to release it.

@phil-davis  let me see how to get it better.

PVince81 commented:
@sharidas sounds good.

I can also have a look if needed as I'm more familiar with this inheritance mess

sharidas committed:
Fix deletion of selected files in trashbin

The file selection process provides the result
which has file name that doesn't match with the
same in the filesystem. This is with reference
to the trashbin. So lets say user tries to selectall,
the value in this.files doesn't give the name
which can be passed to the ajax request to delete
or undelete for later actions. So the purpose
of this change is to check if the selector is
from trashbin, if so adjust the filename as
filename + '.d' + mtime/1000.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

phil-davis committed:
webUI trashbin tests for partial unselection

Signed-off-by: Phil Davis <phil@jankaritech.com>

phil-davis committed:
Unit tests for trashbin partial unselected delete and restore

Signed-off-by: Phil Davis <phil@jankaritech.com>

PVince81 commented on apps/files/js/filelist.js:
```diff
@@ -694,6 +694,16 @@
+					If the selection is from trashbin we need to do update the
+					name of the files.
+					*/
+					var tempDeletedFileName = this.files[i].name + '.d' + this.files[i].mtime/1000;
```
the files app must not know anything about the trashbin app.
here you have added trashbin specific code

PVince81 replied:
normally when using trashbin I expect the "name" attribute to already contain the "d" suffix.
it's "displayName" that contains whatever is visible in the web UI

if your observation was that "name" did not contain the "d" then there's a bug somewhere when populating the "files" array or when reading the "data-file"

PVince81 replied:
looking at: https://github.com/owncloud/core/blob/master/apps/files_trashbin/js/filelist.js#L106

ok so maybe the "files" array doesn't contain the "d" and only the data-file attribute does when creating the row.

So while your code makes sense it's not located in the correct class as it's making the files app depend on trashbin. Maybe the trashbin class must override some of the selection related methods to make it work and inject the missing "d" part when dealing with root entries.

Or introduce new methods in "files" app class and override it in "files_trashbin" when computing the name

sharidas replied:
Looking at the code https://github.com/owncloud/core/blob/master/apps/files/js/filelist.js#L1189 the displayName is already set. Which is why user doesn't see `'.d' + mtime/1000` along with the file names in UI. I still get a feel that it is the names passed during the `delete` and `undelete` ajax requests that need to be altered. And this computation will be done in the 
- https://github.com/owncloud/core/blob/master/apps/files_trashbin/js/filelist.js#L185
- https://github.com/owncloud/core/blob/master/apps/files_trashbin/js/filelist.js#L235
Does this look like back to square one?

sharidas commented on apps/files_trashbin/js/filelist.js:
```diff
@@ -156,6 +156,36 @@
+		/**
+		 * Event handler for when selecting/deselecting all files
+		 */
+		_onClickSelectAll: function(e) {
```
Override selectAll in trashbin. So The only difference between files/js/filelist.js and this filelist.js is from https://github.com/owncloud/core/pull/30993/files#diff-5016b1abbde02ec31bb639507f64d604R170 to https://github.com/owncloud/core/pull/30993/files#diff-5016b1abbde02ec31bb639507f64d604R176
i.e, addition of line:
```
this.files[i].name = this.files[i].name + '.d' +
          Math.floor(this.files[i].mtime/1000);
```

PVince81 replied:
looks better, need to find a way to prevent that much code duplication, could be done by introducing a new smaller method that can be overridden

PVince81 commented on apps/files_trashbin/js/filelist.js:
```diff
@@ -156,6 +156,36 @@
+					name of the files, i.e make the filenames real so that
+					its appended with '.d' + mtime/1000
+					*/
+					this.files[i].name = this.files[i].name + '.d' +
```
not good as this would overwrite the existing data structure contents and will cause side effects

you need to make a copy of the file data, only replace the name there, not in the original array

PVince81 reviewed (changes requested):
seems we're getting closer and now I understand the actual problem better

sharidas commented on apps/files_trashbin/js/filelist.js:
```diff
@@ -156,6 +156,27 @@
+			Any change to trashbinFiles will not have any change to this.files
+			 */
+			var trashbinFiles = [];
+			this.files.forEach((v, i) => {
```
Do we have a better way to achieve a copy of `this.file` update the name or does this look ok? Got this snippet while searching. This change does achieve the goal to not update `this.files`.

PVince81 commented on apps/files/js/filelist.js:
```diff
@@ -704,15 +704,16 @@
 		 * Event handler for when selecting/deselecting all files
 		 */
-		_onClickSelectAll: function(e) {
+		_onClickSelectAll: function(e, modifiedFiles) {
```
Beware, the function is called from a jQuery event and I'm not sure whether the second param could have arbitrary values. Please double check.

Please update the JSDoc of the function

PVince81 reviewed (commented):
Solution looks great, some small tweak, see comment :-)

phil-davis commented:
@PVince81 @sharidas time to merge and backport?
It will be nice to have this little fix in 10.0.9

sharidas commented on apps/files/js/filelist.js:
```diff
@@ -703,16 +703,43 @@
+			 * Check if there are no arbitrary keys that got into
+			 * modifiedFiles, which can cause irregular behaviours.
+			 */
+			if (modifiedFiles && (modifiedFiles.length === this.files.length)) {
```
Tried to verify at least the keys in `modifiedFiles` are same as `this.files`. Hope this check should be sufficient. If the keys are not same then immediate fall back to `this.files` is made at https://github.com/owncloud/core/pull/30993/files#diff-dafefd9519924bec4efa9f5af4af8e68R729

PVince81 replied:
comparing length doesn't guarantee that both arrays aren't the same and this approach looks like a hack

please find a more elegant way to solve this

sharidas commented:
Backport PR: https://github.com/owncloud/core/pull/31700

PVince81 reviewed (approved):
:+1:
